File: solver/demos/arenstorf.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include <iostream>
#include <numbers>
#include <numeric>
#include <valarray>

#include <solver/observer.hpp>
#include <solver/problem.hpp>
#include <solver/runge_kutta.hpp>
#include <solver/solver.hpp>

/*
solve Arenstorf orbit problem
*/

struct arenstorf_model
{
    using state_t = std::valarray<double>;

    double mu;

    arenstorf_model( double m )
        : mu( m )
    {
    }

    state_t
    operator()( double, state_t const& y ) const
    {
        double const y1 = y[0];
        double const y2 = y[1];
        double const y3 = y[2];
        double const y4 = y[3];

        double const r1 = sqrt( ( y1 + mu ) * ( y1 + mu ) + y2 * y2 );
        double const r2 = sqrt( ( y1 - 1 + mu ) * ( y1 - 1 + mu ) + y2 * y2 );

        double const dy1 = y3;
        double const dy2 = y4;
        double const dy3 = y1 + 2 * y4 - ( 1 - mu ) * ( y1 + mu ) / ( r1 * r1 * r1 ) - mu * ( y1 - 1 + mu ) / ( r2 * r2 * r2 );
        double const dy4 = y2 - 2 * y3 - ( 1 - mu ) * y2 / ( r1 * r1 * r1 ) - mu * y2 / ( r2 * r2 * r2 );

        return { dy1, dy2, dy3, dy4 };
    }
};

int
main( int, char** )
{
    std::string const dirname = "arenstorf_data";
    std::string filename;

    using state_t = std::valarray<double>;

    double const tf = 17.0652165601579625588917206249;
    double const dt = 1e-5;

    double const mu = 0.012277471;

    auto arenstorf_pb = ponio::make_problem( arenstorf_model( mu ) );

    state_t const yini = { 0.994, 0., 0., -2.00158510637908252240537862224 };

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk546m.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk54_6m( 1e-5 ), yini, { 0., tf }, dt, observer::file_observer( filename ) );

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk547m.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk54_7m( 1e-5 ), yini, { 0., tf }, dt, observer::file_observer( filename ) );

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk547s.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk54_7s( 1e-5 ), yini, { 0., tf }, dt, observer::file_observer( filename ) );

    return 0;
}

File: ponio/demos/arenstorf.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import dataclasses

import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt

name = "arenstorf"
data_dir = name+"_data"


@dataclasses.dataclass
class simu_data:
    label: str
    data: npt.NDArray[np.float64]

    def time(self):
        return self.data[:, 0]

    def pos_x(self):
        return self.data[:, 1]

    def pos_y(self):
        return self.data[:, 2]

    def vel_x(self):
        return self.data[:, 3]

    def vel_y(self):
        return self.data[:, 4]

    def time_step(self):
        return self.data[:, -1]


def load_data(tagname, label):
    return simu_data(label=label, data=np.loadtxt(os.path.join(data_dir, f"{name}_{tagname}.dat")))


def plot_orbit(data):
    plt.plot(data.pos_x(), data.pos_y(), "-+", label=data.label)


def plot_velocity(data):
    plt.plot(data.vel_x(), data.vel_y(), "-+", label=data.label)


def plot_time_step(data):
    plt.plot(data.time(), data.time_step(), "-+", label=data.label)


os.makedirs(data_dir, exist_ok=True)

make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

rk546m = load_data("rk546m", "RK5(4) 6M")
rk547m = load_data("rk547m", "RK5(4) 7M")
rk547s = load_data("rk547s", "RK5(4) 7S")

# orbit plot
plt.rcParams["figure.figsize"] = (6, 6)
plt.axes().set_aspect('equal', 'box')

plot_orbit(rk546m)
plot_orbit(rk547m)
plot_orbit(rk547s)

plt.axhline(0., 0, 1, linestyle="--", color="grey", linewidth=1)
plt.axvline(0., 0, 1, linestyle="--", color="grey", linewidth=1)

plt.xlim(-1.5, 1.5)
plt.ylim(-1.5, 1.5)

plt.xlabel("$x$")
plt.ylabel("$y$")

plt.tight_layout(pad=1.0)
plt.legend()

plt.savefig("1-arenstorf-orbit_01.png", dpi=200)

plt.title("Arenstorf orbit")
plt.show()

plt.rcParams["figure.figsize"] = plt.rcParamsDefault["figure.figsize"]

# velocity plot
plt.rcParams["figure.figsize"] = (6, 6)
plt.axes().set_aspect('equal', 'box')

plot_velocity(rk546m)
plot_velocity(rk547m)
plot_velocity(rk547s)

plt.axhline(0., 0, 1, linestyle="--", color="grey", linewidth=1)
plt.axvline(0., 0, 1, linestyle="--", color="grey", linewidth=1)

plt.xlim(-2.5, 2.5)
plt.ylim(-3.5, 1.5)

plt.xlabel(r"$\dot{{x}}$")
plt.ylabel(r"$\dot{{y}}$")

plt.tight_layout(pad=1.0)
plt.legend()

plt.savefig("1-arenstorf-orbit_02.png", dpi=200)

plt.title("Arenstorf velocity")
plt.show()

plt.rcParams["figure.figsize"] = plt.rcParamsDefault["figure.figsize"]

# time step history
plot_time_step(rk546m)
plot_time_step(rk547m)
plot_time_step(rk547s)

plt.xlabel(r"$t$")
plt.ylabel(r"$\Delta t$")
plt.legend()

plt.savefig("1-arenstorf-orbit_03.png")

plt.title("Arenstorf orbit, time step size")
plt.show()
